File: code/smap.cpp
#include <iostream>   
#include <fstream>   
#include <string>    
#include <map>    
#include <vector>  
using namespace std;
struct readinfo {
	string reference;
	long long readstart;
	long long readend;
	char dir;
	long long referencestart;
	long long referenceend;
	long long length;
	readinfo(string rf, long long rds, long long rde, char d, long long rfs, long long rfe, long long len) :reference(rf), readstart(rds), readend(rde), dir(d), referencestart(rfs), referenceend(rfe), length(len) {}
};
int main(int argc, char* argv[]) {
	cerr << "-----------start smap----------------" << endl;
	map<string, vector<readinfo>> h_graph;
	if (argc != 3) {
		cerr << "用法: " << argv[0] << " <输入文件路径> <输出文件路径>" << endl;
		return 1;
	}
	string inpath = argv[1];
	string outpath = argv[2];
	ifstream inputfile(inpath);
	if (!inputfile.is_open()) {
		cerr << "无法打开文件: " << inpath << endl;
		return 1; // 退出程序表示错误
	}
	ofstream outputfile(outpath, ios::app);
	if (!outputfile.is_open()) {
		cerr << "无法打开文件: " << outpath << endl;
		return 1; // 退出程序表示错误
	}
	string line;
	long long ans = 0;
	cerr << "开始建图" << endl;
	while (getline(inputfile, line)) {
		ans++;
		if (ans % 1000 == 0) cerr << "已经读取了" << ans << "行数据" << endl;
		string s;
		string rf;
		string rd;
		long long rds, rde, rfs, rfe, len;
		char d;
		int ff = 0;
		bool fff = 0;
		long long num = 0;
		bool f = 0;
		for (long long i = 0; i < line.size(); i++) {
			if (line[i] == '	') {
				num++;
				if (num == 1) {
					auto it = h_graph.find(s);
					if (it != h_graph.end()) f = 1;
				}
				switch (num) {
				case 1:
					rd = s;
					break;
				case 2:
					break;
				case 3:
					rds = stoi(s);
					break;
				case 4:
					rde = stoi(s);
					break;
				case 5:
					d = s[0];
					break;
				case 6:
					rf = s;
					break;
				case 7:
					break;
				case 8:
					rfs = stoi(s);
					break;
				case 9:
					rfe = stoi(s);
					break;
				case 10:
					break;
				case 11:
					break;
				case 12:
					ff = stoi(s);
					break;
				}
				if (num == 12) break;
				s.clear();
			}
			else s += line[i];
			if (num == 12) break;
		}
		if (ff >= 50) {
			len = rde - rds;
			if (f == 0) {
				h_graph.insert({ rd , { readinfo(rf,rds,rde,d,rfs,rfe,len) } });
				//if (rf == "chr_8_contig_7457") cout << 1 << endl;
			}
			else {
				//if (rf == "chr_8_contig_7457") cout << 1 << endl;
				for (auto& ri : h_graph[rd]) {
					if (ri.reference == rf) {
						ri.length += len;
						fff = 1;
						break;
					}
				}
				if (fff == 0) {
					h_graph[rd].push_back(readinfo(rf, rds, rde, d, rfs, rfe, len));
				}
			}
		}
	}
	cerr << "建图完成" << endl;
	cerr << "开始写入" << endl;
	for (const auto& ele : h_graph) {
		string read = ele.first;
		outputfile << read << ':' << endl;
		//cout << read << endl;
		const vector<readinfo>& rinfo = ele.second;
		for (const readinfo& info : rinfo) {
			//cout << info.length << endl;
			//if (read == "000e5f4e-dc26-4838-b4a0-6badd612a002" && info.reference == "chr_4_contig_3490") cout << 1 << endl;
			if (info.length >= 100) {
				outputfile << info.reference << ',' << info.readstart << ',' << info.readend << ',' << info.dir << ',' << info.referencestart << ',' << info.referenceend << ',' << info.length << endl;
			} 
		}
	}
	cerr << "写入完成" << endl;
	inputfile.close();
	outputfile.close();
	return 0;
}

File: code/smap.py
import sys
from collections import defaultdict

class ReadInfo:
    __slots__ = ('reference', 'readstart', 'readend', 'dir', 'referencestart', 'referenceend', 'length')
    
    def __init__(self, rf, rds, rde, d, rfs, rfe, len_val):
        self.reference = rf
        self.readstart = rds
        self.readend = rde
        self.dir = d
        self.referencestart = rfs
        self.referenceend = rfe
        self.length = len_val

def main():
    # 验证命令行参数
    if len(sys.argv) != 3:
        sys.stderr.write(f"用法: {sys.argv[0]} <输入文件路径> <输出文件路径>\n")
        sys.exit(1)
    
    inpath = sys.argv[1]
    outpath = sys.argv[2]
    
    sys.stderr.write("----------- PAF文件处理开始 -----------\n")
    
    # 打开输入文件
    try:
        inputfile = open(inpath, 'r')
    except IOError:
        sys.stderr.write(f"错误: 无法打开输入文件: {inpath}\n")
        sys.exit(1)
    
    # 打开输出文件（覆盖模式）
    try:
        outputfile = open(outpath, 'w')
    except IOError:
        sys.stderr.write(f"错误: 无法打开输出文件: {outpath}\n")
        sys.exit(1)
    
    # 用于存储每个read的比对信息
    read_data = defaultdict(list)
    line_count = 0
    filtered_count = 0
    
    sys.stderr.write("开始处理PAF文件...\n")
    
    for line in inputfile:
        line_count += 1
        if line_count % 10000 == 0:
            sys.stderr.write(f"已处理 {line_count} 行，过滤 {filtered_count} 条低质量比对\n")
        
        tokens = line.strip().split('\t')
        # 跳过格式不正确的行
        if len(tokens) < 12:
            continue
            
        read_id = tokens[0]
        try:
            # 解析比对信息
            read_start = int(tokens[2])
            read_end = int(tokens[3])
            direction = tokens[4][0] if tokens[4] else ''
            reference = tokens[5]
            ref_start = int(tokens[7])
            ref_end = int(tokens[8])
            mapping_quality = int(tokens[11])
        except (ValueError, IndexError) as e:
            sys.stderr.write(f"警告: 第 {line_count} 行解析错误: {str(e)}\n")
            continue
        
        # 过滤低质量比对
        if mapping_quality < 50:
            filtered_count += 1
            continue
        
        # 计算比对长度
        alignment_length = read_end - read_start
        
        # 检查是否已存在相同参考序列的记录
        found = False
        for existing in read_data[read_id]:
            if existing.reference == reference:
                existing.length += alignment_length
                found = True
                break
        
        # 若不存在则添加新记录
        if not found:
            read_data[read_id].append(
                ReadInfo(reference, read_start, read_end, direction, 
                         ref_start, ref_end, alignment_length)
            )
    
    sys.stderr.write(f"PAF文件处理完成，总行数: {line_count}, 过滤低质量比对: {filtered_count}\n")
    sys.stderr.write(f"有效比对记录数: {sum(len(v) for v in read_data.values())}\n")
    
    # 写入输出文件
    sys.stderr.write("开始写入比对摘要...\n")
    output_records = 0
    
    for read_id, entries in read_data.items():
        outputfile.write(f"{read_id}:\n")
        for info in entries:
            # 只保留长度≥100bp的比对
            if info.length >= 100:
                output_line = (
                    f"{info.reference},{info.readstart},{info.readend},"
                    f"{info.dir},{info.referencestart},{info.referenceend},{info.length}\n"
                )
                outputfile.write(output_line)
                output_records += 1
    
    sys.stderr.write(f"写入完成，输出记录数: {output_records}\n")
    inputfile.close()
    outputfile.close()
    sys.stderr.write("----------- PAF文件处理结束 -----------\n")

if __name__ == "__main__":
    main()